Modules/posixmodule.c

```diff
@@ -6737,6 +6739,96 @@ os_register_at_fork_impl(PyObject *module, PyObject *before,
+#if defined(__APPLE__) && defined(HAVE_GETPID)
+    mach_port_t macos_self = mach_task_self();
+    mach_port_t macos_task;
+    if (task_for_pid(macos_self, getpid(), &macos_task) == KERN_SUCCESS) {
+        thread_array_t macos_threads;
+        mach_msg_type_number_t macos_n_threads;
+        if (task_threads(macos_task, &macos_threads,
+                         &macos_n_threads) == KERN_SUCCESS) {
+            num_python_threads = macos_n_threads;
+        }
+    }
+#elif defined(__linux__)
+    // Linux /proc/self/stat 20th field is the number of threads.
+    FILE* proc_stat = fopen("/proc/self/stat", "r");
+    if (proc_stat) {
+        size_t n;
+        char stat_line[160];
+        n = fread(&stat_line, 1, 159, proc_stat);
+        stat_line[n] = '\0';
+        fclose(proc_stat);
+
+        char *saveptr = NULL;
+        char *field = strtok_r(stat_line, " ", &saveptr);
+        unsigned int idx;
+        for (idx = 19; idx && field; --idx) {
+            field = strtok_r(NULL, " ", &saveptr);
+        }
+        if (idx == 0 && field) {  // found the 20th field
+            num_python_threads = atoi(field);  // 0 on error
+        }
+    }
+#endif
+    if (num_python_threads <= 0) {
+        // Fall back to just the number of threads this CPython runtime knows about.
+        PyObject *threading = PyImport_GetModule(&_Py_ID(threading));
+        if (!threading) {
+            PyErr_Clear();
+            return;
+        }
+        PyObject *threading_active = PyObject_GetAttr(threading, &_Py_ID(_active));
+        if (!threading_active) {
+            PyErr_Clear();
+            Py_DECREF(threading);
+            return;
+        }
+        PyObject *threading_limbo = PyObject_GetAttr(threading, &_Py_ID(_limbo));
+        if (!threading_limbo) {
+            PyErr_Clear();
+            Py_XDECREF(threading);
+            Py_XDECREF(threading_active);
+            return;
+        }
+        // Worst case if someone replaced threading._active or threading._limbo
+        // with non-dicts, we get -1 from *Length() below and undercount.
+        // Whatever. That shouldn't happen, we're best effort so we clear errors
+        // and move on.
+        assert(PyMapping_Check(threading_active));
+        assert(PyMapping_Check(threading_limbo));
+        // Duplicating what threading.active_count() does but without holding
+        // threading._active_limbo_lock so our count could be inaccurate if these
+        // dicts are mid-update from another thread.  Not a big deal.
+        num_python_threads = (PyMapping_Length(threading_active)
+                              + PyMapping_Length(threading_limbo));
+        PyErr_Clear();
+        Py_DECREF(threading);
+        Py_DECREF(threading_active);
+        Py_DECREF(threading_limbo);
+    }
+    if (num_python_threads > 1) {
+        PyErr_WarnFormat(
+                PyExc_DeprecationWarning, 1,
+                "multi-threaded process, %s() may cause deadlocks.", name);
```

Comment:
Should we point to more information to understand this issue somewhere, like a FAQ entry? I know we don't usually do this but this is somewhat of an esoteric issue, might be deep inside a library (e.g. something using multiprocessing), and we _have_ been trying to improve error messages to be more helpful...

Reply:
I don't have a good canonical link to propose right now. The other thing that often happens is that the error message winds up with a Google search result of a stackoverflow or similar live content user moderated question linking to the current best sources of info.

For multiprocessing, I intend to do the warning about start_method `fork` as the default being used. In that specific default-fork situation where multiprocessing will provide its own DeprecationWarning, it probably makes sense to have multiprocessing silence this one on its call.

Reply:
Wording improved to include the pid and indicate that deadlocks could happen in the child process. Searching, I don't see a canonical online reference to the fork+threads issue of high enough quality and domain owner stability that I'd want to codify it within our error message. So I'll leave this without a URL for now, we can add one later if desired.

If we want one, we probably need to write a little section covering the topic in our docs somewhere and link to that.